### PR TITLE
Create API endpoint for articles

### PR DESCRIPTION
**Pull request by danfrenette**
Thanks a ton to anyone reviewing this!

This change adds the blueprinter gem to create a JSON API endpoint to
retreive the published articles in the database. This gem was chosen
over Active Model Serializers because that gem is not currently
maintained.

The endpoint exposes an articles id, title, description, tags, and published date.

This PR is also the base branch for the React side of this implementation, so feel free to check that out: https://github.com/danfrenette/a-different-shade-of-blue/pull/7

### CODE REVIEW

File: spec/models/article_spec.rb [hunk 1]
```diff
@@ -16,6 +16,22 @@
     it { should validate_uniqueness_of :title }
   end
 
+  describe ".published" do
+    it "returns the articles that have a published_at timestamp before the current time" do
+      published_article = create :article
+      article_published_today = create :article, published_at: 3.hours.ago
+      _unpublished_article = create :article, :unpublished
+      _article_to_be_published_today = create :article, published_at: 3.hours.from_now
```
Comment (mlodato517) [inline spec/models/article_spec.rb:24]: 🎱 This could be super picky so feel free to ignore but this test raises a few questions for me:
1. is it a good idea to rely on the default `create :article` to create an article that is already published? Might be better to make that explicit so if the base factory changes this test still passes
1. is there a reason for `3.hours`? Doesn't seem like it from the scope but this test  makes it seem like that's the magic number
1. is there a reason to have two published and two unpublished items? Is it proving that you don't just get the single most recent published item? If so, is there a reason to have two unpublished items?
1. using the `:unpublished` trait here is a _little_ suspicious (to me and probably just me). It's almost tautological - it's like this test is testing that articles with the `:unpublished` trait aren't included in the result. It also makes me wonder why there isn't a trait for `:published`. But then I also feel like a test like:
```rb
published = create :article, :published
_unpublished = create :article, :unpublished
expect(Article.Published).to match_array([published])
```
tells the reader not much about what the scope does - just that it works with the traits and you have to go inspect the traits or scope to figure out what it's supposed to be doing? Or I guess just trust the test title. Which is maybe 👌

File: Gemfile [hunk 2]
```diff
@@ -5,6 +5,7 @@ ruby '2.6.5'
 gem 'rails', '~> 6.0.3.1'
 
 gem 'acts-as-taggable-on'
+gem 'blueprinter'
```
Comment (kevin-j-m) [inline Gemfile:8]: Given the name of the repo, this seems an appropriate choice.

File: spec/models/article_spec.rb [hunk 3]
```diff
@@ -16,6 +16,22 @@
...
+
+      expect(Article.published.ids).to eq(
+        [
+          published_article.id,
+          article_published_today.id,
+        ],
+      )
```
Comment (mlodato517) [inline spec/models/article_spec.rb:31]:
```suggestion
      expect(Article.published).to match_array(
        [
          published_article,
          article_published_today,
        ],
      )
```
1. Swapped `eq` to `match_array` so we don't care about order (unless you _do_ care about order in which case the scope can be ordered) (also, I _think_ that's the right matcher ... jeez I haven't done rails in so long)
1. Removed the `ids` for the full query - less performant so feel free to ignore but also I'm not sure about the balance of readability in specs vs performance. And I imagine when querying a table with 4 records in it, it should be pretty fast. Unless the records are huge. In which case maybe the base factory should create smaller records. But also, feel free to ignore this and leave the ids in!

File: app/controllers/api/articles_controller.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+module Api
+  class ArticlesController < ApplicationController
+    def index
+      render json: ArticleBlueprint.render(Article.published)
```
Comment (kevin-j-m) [inline app/controllers/api/articles_controller.rb:4]: Do you want any pagination here?

File: spec/requests/articles_spec.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      get api_articles_path
+
+      expect(json_response).to contain_exactly(
+        {
+          "id" => published_article.id,
+          "title" => "Published Article",
+          "description" => "Awesome description",
+          "publishedAt" => "01/01/2000",
+          "tags" => a_collection_containing_exactly(
+            a_hash_including("name" => "educational"),
```
Comment (mlodato517) [inline spec/requests/articles_spec.rb:22]: I assume these tags are related to the `:with_tags` trait? If so, that might need to be explicit. I think traits are nice for when you need something of a certain type so that some validation passes and you can test the _real_ thing you care about. But if you're testing the guts of a trait, then if you leave them out of the test the reader has to go check the trait to verify these tags. And then if the trait changes this test fails when, in theory, it shouldn't.

File: app/serializers/article_blueprint.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+class ArticleBlueprint < Blueprinter::Base
+  identifier :id
+
+  fields :title, :description
+  field :published_at, datetime_format: "%m/%d/%Y", name: :publishedAt
```
Comment (kevin-j-m) [inline app/serializers/article_blueprint.rb:5]: Why the change in casing convention for the name?

File: spec/models/article_spec.rb [hunk 7]
```diff
@@ -16,6 +16,22 @@
     it { should validate_uniqueness_of :title }
   end
 
+  describe ".published" do
+    it "returns the articles that have a published_at timestamp before the current time" do
+      published_article = create :article
```
Comment (kevin-j-m) [inline spec/models/article_spec.rb:21]: If this is intended to be the current time then:
1. I would explicitly set `published_at` to be now. It's not clear if the factory or model or anything defines a default, and the value of that attribute is important/relevant to this spec.
2. I would freeze time so that this stays definitively the same moment in time when `Article.published` is called.

File: spec/models/article_spec.rb (hunk 3 above)
Comment (kevin-j-m) [inline spec/models/article_spec.rb:31]: Why this over

```ruby
expect(Article.published).to match_array [published_article, article_published_today]
```

File: spec/requests/articles_spec.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
+require "rails_helper"
+
+RSpec.describe "Articles API", type: :request do
```
Comment (mlodato517) [inline spec/requests/articles_spec.rb:3]: 🎱 
```suggestion
RSpec.describe "Articles API" do
```
It looks like your tests are set up to [infer the test type](https://github.com/danfrenette/a-different-shade-of-blue/blob/develop/spec/rails_helper.rb#L71). If you're going to specify them explicitly everywhere I'd remove them from the `rails_helper`. If you're going to leave the config in, I wouldn't specify the test unless it's in an "unexpected" place.

File: spec/models/article_spec.rb [hunk 9]
```diff
@@ -16,6 +16,22 @@
...
+      article_published_today = create :article, published_at: 3.hours.ago
+      _unpublished_article = create :article, :unpublished
+      _article_to_be_published_today = create :article, published_at: 3.hours.from_now
+
+      expect(Article.published.ids).to eq(
```
Comment (kevin-j-m) [inline spec/models/article_spec.rb:26]:
```suggestion
      expect(Article.published.ids).to match_array(
```

It doesn't look like this scope defines ordering, so you can't guarantee items will be in this order.

Comment (kevin-j-m) [reply spec/models/article_spec.rb:26]: If that's important/expected, I would recommend changing the scope to set an order.

File: spec/requests/articles_spec.rb [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  describe "GET /articles" do
+    it "returns the published articles" do
+      published_article = create :article,
+        :with_tags,
```
Comment (kevin-j-m) [inline spec/requests/articles_spec.rb:7]: Given this, I don't know what the tags are, so it's not clear why those particular tags show up in the response.

File: spec/requests/articles_spec.rb [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+        title: "Published Article",
+        description: "Awesome description",
+        published_at: Date.new(2000, 1, 1)
+      _unpublished_article = create :article, :unpublished, title: "Unpublished Article"
+
+      get api_articles_path
+
+      expect(json_response).to contain_exactly(
+        {
+          "id" => published_article.id,
```
Comment (mlodato517) [inline spec/requests/articles_spec.rb:17]: Is the rocket syntax needed for this hash because JSON?

File: spec/support/request_macros.rb [hunk 12]
```diff
@@ -0,0 +1,5 @@
+module RequestMacros
+  def json_response
```
Comment (mlodato517) [inline spec/support/request_macros.rb:2]: TIL! This is pretty slick. Definitely cleans stuff up

File: Gemfile (hunk 2 above)
Comment (danfrenette) [reply Gemfile:8]: I'm incredibly organized like that.

File: spec/models/article_spec.rb (hunk 1 above)
Comment (danfrenette) [reply spec/models/article_spec.rb]: 1. Agreed, changed
2. The scope is testing articles that are published, meaning they have a `published_at` timestamp set in the future. I think my thought process was just that I wanted to make sure this took time into consideration and not just the date, hence why there's another article set to be published three hours from now, but this might be confusing / testing the wrong thing.
3. I think my response above answers this as well
4. The convention I was following is that `create :article` creates an article record without having to specify any traits or set any values. However, `published_at` is a null false column, because I don't want any articles to exist in the database that will never be published (kind of rethinking that now actually). I think the best compromise moving forward is to explicitly set the times on just two article records without using traits to be as verbose as possible.

File: spec/models/article_spec.rb (hunk 9 above)
Comment (danfrenette) [reply spec/models/article_spec.rb]: I think when the order becomes a concern, I'll add a separate scope for that, but this is fine for now.

File: spec/requests/articles_spec.rb (hunk 8 above)
Comment (danfrenette) [reply spec/requests/articles_spec.rb:3]: Made a ticket to remove these in a separate unit of work since they're in every spec as of right now.

File: spec/requests/articles_spec.rb (hunk 5 above)
Comment (danfrenette) [reply spec/requests/articles_spec.rb]: Added one explicitly 👍   unfortunately it's through the acts-as-taggable-on gem, so the way to do that still isn't super clear, but I think it's still a step in the right direction.
